zebra-chain/src/commitments/sapling.rs

```diff
@@ -0,0 +1,341 @@
+        let pk_d_bytes = <[u8; 32]>::from(transmission_key);
+        let v_bytes = value.to_bytes();
+
+        s.append(&mut BitVec::<Lsb0, u8>::from_slice(&g_d_bytes[..]));
+        s.append(&mut BitVec::<Lsb0, u8>::from_slice(&pk_d_bytes[..]));
+        s.append(&mut BitVec::<Lsb0, u8>::from_slice(&v_bytes[..]));
+
+        let rcm = CommitmentRandomness(generate_trapdoor(csprng));
+
+        (
+            rcm,
+            NoteCommitment::from(windowed_pedersen_commitment(rcm.0, &s)),
+        )
+    }
+
+    /// Hash Extractor for Jubjub (?)
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#concreteextractorjubjub
+    pub fn extract_u(&self) -> jubjub::Fq {
+        self.0.get_u()
+    }
+}
+
+/// A Homomorphic Pedersen commitment to the value of a note, used in Spend and
+/// Output Descriptions.
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretehomomorphiccommit
+#[derive(Clone, Deserialize, PartialEq, Serialize)]
+pub struct ValueCommitment(#[serde(with = "serde_helpers::AffinePoint")] pub jubjub::AffinePoint);
+
+impl fmt::Debug for ValueCommitment {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("ValueCommitment")
+            .field("u", &hex::encode(self.0.get_u().to_bytes()))
+            .field("v", &hex::encode(self.0.get_v().to_bytes()))
+            .finish()
+    }
+}
+
+impl From<[u8; 32]> for ValueCommitment {
+    fn from(bytes: [u8; 32]) -> Self {
+        Self(jubjub::AffinePoint::from_bytes(bytes).unwrap())
+    }
+}
+
+impl From<jubjub::ExtendedPoint> for ValueCommitment {
+    fn from(extended_point: jubjub::ExtendedPoint) -> Self {
+        Self(jubjub::AffinePoint::from(extended_point))
+    }
+}
+
+impl Eq for ValueCommitment {}
+
+impl From<ValueCommitment> for [u8; 32] {
+    fn from(cm: ValueCommitment) -> [u8; 32] {
+        cm.0.to_bytes()
+    }
+}
+
+/// LEBS2OSP256(repr_J(cv))
+///
+/// https://zips.z.cash/protocol/protocol.pdf#spendencoding
+/// https://zips.z.cash/protocol/protocol.pdf#jubjub
+impl ZcashSerialize for ValueCommitment {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_all(&self.0.to_bytes())?;
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for ValueCommitment {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        Ok(Self(
+            jubjub::AffinePoint::from_bytes(reader.read_32_bytes()?).unwrap(),
+        ))
+    }
+}
+
+impl ValueCommitment {
+    /// Generate a new _ValueCommitment_.
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#concretehomomorphiccommit
+    // TODO: accept an Amount instead?
+    #[allow(non_snake_case)]
+    pub fn new<T>(csprng: &mut T, value_bytes: [u8; 32]) -> Self
+    where
+        T: RngCore + CryptoRng,
+    {
+        let v = jubjub::Fr::from_bytes(&value_bytes).unwrap();
+        let rcv = generate_trapdoor(csprng);
+
+        let V = find_group_hash(*b"Zcash_cv", b"v");
```

Comment:
in the future, a nice optimization might be to keep the pedersen generators as constants to avoid having to recompute them on every commitment call, but I don't think we should worry about this now.

Reply:
Agreed, they are repeated, but are so short and clear that I didn't worry too much.

Reply:
Well, computing these night do multiple GroupHashes (I don't think I've ever checked how many it does).

Reply:
I think it would be fine to leave a note that this could be optimized but not actually do anything about it, since we're just validating existing amounts in our first milestone I expect that this code won't actually be called until we implement more features.